src/script/interpreter.cpp

```diff
@@ -1264,6 +1280,116 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                        return set_error(serror, *err);
+                    }
+
+                    popstack(stack);
+                    popstack(stack);
+                    stack.push_back(vchTrue);
+                }
+                break;
+
+                case OP_VAULT:
+                {
+                    // Only available in Tapscript
+                    if (sigversion == SigVersion::BASE || sigversion == SigVersion::WITNESS_V0) {
+                        return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
+                    }
+
+                    // Stack:
+                    //  - <flu-script>
+                    //  - <n-pushes>
+                    //  [ n items ... ]
+                    //  - <trigger-vout-idx>
+                    //  - <revault-vout-idx>
+                    //  - <revault-amount>
+                    if (stack.size() < 5) {
+                        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+                    }
+
+                    const valtype& flu_script_data{stacktop(-1)};
+                    CScript flu_script_body{flu_script_data.begin(), flu_script_data.end()};
+                    popstack(stack);
+
+                    const CScriptNum n_pushes_data{stacktop(-1), fRequireMinimal};
+                    const size_t n_pushes = static_cast<size_t>(n_pushes_data.GetInt64());
+                    popstack(stack);
+
+                    // Needed on stack: n pushes + the 2 vout indices
+                    if (stack.size() < (n_pushes + 2)) {
+                        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+                    }
+
+                    // Extract the `n` data pushes and construct the entire expected
+                    // FORWARD_LEAF_UPDATE script.
+                    std::vector<valtype> flu_data{
+                        std::make_move_iterator(stack.end() - n_pushes),
+                        std::make_move_iterator(stack.end())};
+                    stack.erase(stack.end() - n_pushes, stack.end());
+
+                    CScript flu_script = PushAll(flu_data);
+                    flu_script.reserve(flu_script.size() + flu_script_body.size());
+                    move_to_end(flu_script, flu_script_body);
+
+                    // Indicates which of the vouts carries forward
+                    // the value from the vault into the unvault trigger output.
+                    std::optional<size_t> trigger_vout_idx = GetVoutIdxFromStack(-1);
+                    if (!trigger_vout_idx) {
+                        return set_error(serror, SCRIPT_ERR_VAULT_BAD_VOUT_IDX);
+                    }
+
+                    // Indicates which (if any) of the vouts is a "revault" of some
+                    // portion of the value back into the original vault sPK.
+                    std::optional<size_t> revault_vout_idx = GetVoutIdxFromStack(-2);
```

Comment:
Feedback for the BIP. I think this is a malleability vector if out < 0. I can see a transaction in the fly and set the value to `int32::MIN` decreasing the fee rate of the transaction. I think we should force that this should be -1 no vout is present.

Reply:
Fixed.